modules/Bio/EnsEMBL/IO/Writer/VCF4.pm

```diff
@@ -0,0 +1,187 @@
+=head1 LICENSE
+
+Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+Copyright [2016] EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=head1 NAME
+
+Bio::EnsEMBL::IO::Writer::VCF4 - Generic VCF4 Writer
+
+=head1 SYNOPSIS
+
+  use Bio::EnsEMBL::IO::Writer::VCF4;
+
+  my $writer = Bio::EnsEMBL::IO::Writer::VCF4->new($translator);
+
+  $writer->write($object);
+
+=head1 Description
+
+  Write records out in VCF4 format. The module uses a translator given
+  at creation time which knows how to interrogate a specified type
+  of feature/object that will be passed in to the write function.
+
+  As in, for the VCF4 format the translator must know how to retrieve
+  CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT and individual
+  genotypes from he objects you plan to feed the writer.
+
+=cut
+
+package Bio::EnsEMBL::IO::Writer::VCF4;
+
+use base qw/Bio::EnsEMBL::IO::Writer::ColumnBasedGeneric/;
+
+use strict;
+use warnings;
+use Carp;
+
+use Bio::EnsEMBL::IO::Object::VCF4;
+use Bio::EnsEMBL::IO::Object::VCF4Metadata;
+
+=head2 new
+
+  Description: Constructor for a column based generic writer
+  Args[1]  : Translator object for the type of object being written 
+         (ie. for Ensembl Features, etc)
+
+=cut
+
+sub new {
+  my $class = shift;
+  my $translator = shift;
+
+  my $self = $class->SUPER::new($translator);
+
+  $self->fields(Bio::EnsEMBL::IO::Object::VCF4->fields());
+
+  # Cheat and make a VCF4Metadata object in a really quick and lighttweight manner.
+  # We need one hanging around for the fwd-ref writer call
+  #$self->{directive} = bless { type => 'directive' }, 'Bio::EnsEMBL::IO::Object::VCF4Metadata';
+  #$self->{header} = bless { type => 'header' }, 'Bio::EnsEMBL::IO::Object::VCF4Metadata';
+
+  return $self;
+}
+
+#=head2 header
+#
+#  Description: Shortcut to write the header in the VCF4 format,
+#         ie. #CHROM POS ID REF ALT QUAL FILTER INFO FORMAT
+#
+#=cut
+
+#sub header {
+#  my $self = shift;
+#
+#  $self->write( $self->{header} );
+#}
+
+=head2 create_record
+
+  Description: Create the record in VCF4 to write out to the file
+  Args[1]  : Object to format
+  Returntype : String
+
+=cut
+
+sub create_record {
+  my $self = shift;
+  my $object = shift;
+
+  my $fields = $self->fields();
+  my @values = $self->{translator}->batch_fields($object, $fields);
```

Comment:
Writers have a $self->translator getter which should be used instead of reaching into the object internals.